python/cm/adcm_config.py

```diff
@@ -342,33 +342,94 @@ def group_is_activatable(spec):
     return False
 
 
-def get_builtin_variant(obj, func_name):
-    def free_host(obj):
-        out = []
-        for host in Host.objects.filter(cluster=None):
-            out.append(host.fqdn)
-        return out
+def get_cluster(obj):
+    if obj.prototype.type == 'service':
+        cluster = obj.cluster
+    elif obj.prototype.type == 'host':
+        cluster = obj.cluster
+    elif obj.prototype.type == 'cluster':
+        cluster = obj
+    else:
+        return None
+    return cluster
 
-    def cluster_host(obj):
-        out = []
-        if obj.prototype.type == 'service':
-            cluster = obj.cluster
-        elif obj.prototype.type == 'cluster':
-            cluster = obj
-        else:
+
+def variant_service_in_cluster(obj, args=None):
+    out = []
+    cluster = get_cluster(obj)
+    if not cluster:
+        return []
+
+    for co in ClusterObject.objects.filter(cluster=cluster):
+        out.append(co.prototype.name)
+    return out
+
+
+def variant_service_to_add(obj, args=None):
+    out = []
+    cluster = get_cluster(obj)
+    if not cluster:
+        return []
+
+    for proto in Prototype.objects.filter(bundle=cluster.prototype.bundle, type='service'):
+        if ClusterObject.objects.filter(prototype=proto, cluster=cluster):
+            continue
+        out.append(proto.name)
+    return out
+
+
+def variant_host_in_cluster(obj, args):
+    out = []
+    cluster = get_cluster(obj)
+    if not cluster:
+        return []
+
+    if 'service' in args:
```

Comment:
if args = None, then there will be TypeError: argument of type 'NoneType' is not iterable

Reply:
thanks!